File: paipan/paipan/china_dst.py
"""
China DST correction. Port of paipan-engine/src/chinaDst.js.
Only 1986-05-04 ~ 1991-09-15 summers had DST in China.

During DST the wall clock was advanced +1h, so to recover real (standard)
time we subtract 1 hour from the clock reading.
"""
from __future__ import annotations
from datetime import datetime, timedelta, timezone
from zoneinfo import ZoneInfo

_HOST_TZ = ZoneInfo("Asia/Shanghai")


# NOTE: chinaDst.js:16-24 — CHINA_DST_PERIODS 照抄 Node 源码数据。
# 格式：{year: (start_month, start_day, end_month, end_day)}
# 起止边界钟点统一在 02:00，由 _is_in_dst 施加。
_DST_TABLE: dict[int, tuple[int, int, int, int]] = {
    1986: (5, 4, 9, 14),   # NOTE: chinaDst.js:18
    1987: (4, 12, 9, 13),  # NOTE: chinaDst.js:19
    1988: (4, 10, 9, 11),  # NOTE: chinaDst.js:20
    1989: (4, 16, 9, 17),  # NOTE: chinaDst.js:21
    1990: (4, 15, 9, 16),  # NOTE: chinaDst.js:22
    1991: (4, 14, 9, 15),  # NOTE: chinaDst.js:23
}


def _is_in_dst(year: int, month: int, day: int, hour: int) -> bool:
    """Port of chinaDst.js:34 isChinaDst.

    Node 用 `ts >= startTs && ts < endTs`，起止日都在 02:00。
    这里 minute 参数 Node 也没用（只比较到 hour），为忠实起见保持同样行为。
    """
    entry = _DST_TABLE.get(year)
    if entry is None:
        return False
    sm, sd, em, ed = entry
    # NOTE: chinaDst.js:38 — Node 构造时 minute=0, second=0，按 hour 粒度比较
    t = datetime(year, month, day, hour, 0, 0)
    # NOTE: chinaDst.js:40 — 起始：起始日 02:00
    start = datetime(year, sm, sd, 2, 0, 0)
    # NOTE: chinaDst.js:42 — 结束：结束日 02:00（含该日 00:00-02:00）
    end = datetime(year, em, ed, 2, 0, 0)
    # NOTE: chinaDst.js:44 — ts >= startTs && ts < endTs
    return start <= t < end


def correct_china_dst(
    year: int, month: int, day: int, hour: int, minute: int
) -> dict:
    """Port of chinaDst.js:51 correctChinaDst.

    Returns:
        {year, month, day, hour, minute, wasDst}
    """
    # NOTE: chinaDst.js:52 — isChinaDst 只吃 hour，不吃 minute
    in_dst = _is_in_dst(year, month, day, hour)
    if not in_dst:
        # NOTE: chinaDst.js:54 — 字段顺序及 camelCase wasDst 照抄
        return {
            "year": year,
            "month": month,
            "day": day,
            "hour": hour,
            "minute": minute,
            "wasDst": False,
        }
    # NOTE: chinaDst.js:56 — new Date(y, m-1, d, hour-1, minute, 0)
    # JS Date constructor under host Asia/Shanghai does "wall-clock construction
    # with forward-normalize on DST gap" — uses IANA tzdata. If hour-1 lands in
    # a non-existent wall-clock (spring-forward), JS normalizes forward.
    # Critically, IANA's 1988 CDT start is 1988-04-17, not Node table's 04-10,
    # so JS's behavior varies per year. Replicate via zoneinfo UTC round-trip.
    wc = (
        datetime(year, month, day, hour - 1, minute, 0, tzinfo=_HOST_TZ)
        .astimezone(timezone.utc)
        .astimezone(_HOST_TZ)
    )
    return {
        "year": wc.year,
        "month": wc.month,
        "day": wc.day,
        "hour": wc.hour,
        "minute": wc.minute,
        "wasDst": True,
    }

File: archive/paipan-engine/src/chinaDst.js
/**
 * 中国 1986-1991 年夏令时修正
 *
 * 中国从 1986 年至 1991 年实行过 6 年夏令时（简称 DST）。
 * 每年 4-5 月的某个周日 02:00 起拨快 1 小时，
 * 9 月的某个周日 02:00 起拨回 1 小时。
 *
 * 处于 DST 期间的出生时间，钟表显示时比真实太阳钟早 1 小时，
 * 排八字时必须先减 1 小时再做后续处理。
 *
 * 数据来源：国务院 1986 年 4 月 12 日发布的关于实行夏令时的通知，
 * 及国务院 1992 年 4 月 5 日发布的关于停止实行夏令时的通知。
 */

// 每年 DST 的起止日期（闭区间起，闭区间止 00:00 - 不含结束日 02:00 之后）
const CHINA_DST_PERIODS = [
  // 年份, 开始 [月,日], 结束 [月,日]
  { year: 1986, start: [5, 4],  end: [9, 14] },
  { year: 1987, start: [4, 12], end: [9, 13] },
  { year: 1988, start: [4, 10], end: [9, 11] },
  { year: 1989, start: [4, 16], end: [9, 17] },
  { year: 1990, start: [4, 15], end: [9, 16] },
  { year: 1991, start: [4, 14], end: [9, 15] },
];

/**
 * 判断指定的"钟表显示时间"是否处于中国夏令时期间
 * @param {number} year
 * @param {number} month - 1-12
 * @param {number} day
 * @param {number} hour - 0-23
 * @returns {boolean}
 */
function isChinaDst(year, month, day, hour) {
  const period = CHINA_DST_PERIODS.find((p) => p.year === year);
  if (!period) return false;

  const ts = new Date(year, month - 1, day, hour, 0, 0).getTime();
  // 开始：起始日 02:00
  const startTs = new Date(year, period.start[0] - 1, period.start[1], 2, 0, 0).getTime();
  // 结束：结束日 02:00（含该日 00:00-02:00）
  const endTs = new Date(year, period.end[0] - 1, period.end[1], 2, 0, 0).getTime();

  return ts >= startTs && ts < endTs;
}

/**
 * 如果在 DST 期间，返回真实太阳时对应的钟表时间（减 1 小时）
 * @returns {{year, month, day, hour, minute, wasDst}}
 */
function correctChinaDst(year, month, day, hour, minute) {
  const inDst = isChinaDst(year, month, day, hour);
  if (!inDst) {
    return { year, month, day, hour, minute, wasDst: false };
  }
  const d = new Date(year, month - 1, day, hour - 1, minute, 0);
  return {
    year: d.getFullYear(),
    month: d.getMonth() + 1,
    day: d.getDate(),
    hour: d.getHours(),
    minute: d.getMinutes(),
    wasDst: true,
  };
}

module.exports = { isChinaDst, correctChinaDst, CHINA_DST_PERIODS };
